tests/run_tests.sh

```diff
@@ -0,0 +1,351 @@
+NOT_OUTPUTS=""
+touch ${TEST_OUTPUT}/${TEST}_unmatched-read-1.out.gz
+$BARCODE_SPLITTER --mismatches 2 --bcfile "${TEST_DATA}/barcode_splitter_barcodes.txt" "${TEST_DATA}/barcode_splitter1.fastq" --prefix "${TEST_OUTPUT}/${TEST}_" --suffix .out --idxread 1 --split_all --gzipout $DEBUG_OPT 2> ${TEST_OUTPUT}/${TEST}_error.out 1> ${TEST_OUTPUT}/${TEST}_summary.out
+EXIT_CODE=$?
+EXPECTED_EXIT_CODE=9
+check_test_results
+
+
+# Test gzipped output sample file exists
+TEST=test_20
+rm -f ${TEST_OUTPUT}/${TEST}_* 2> /dev/null
+OUTPUTS="summary.out error.out"
+NOT_OUTPUTS=""
+touch ${TEST_OUTPUT}/${TEST}_BC1-read-1.out.gz
+$BARCODE_SPLITTER --mismatches 2 --bcfile "${TEST_DATA}/barcode_splitter_barcodes.txt" "${TEST_DATA}/barcode_splitter1.fastq" --prefix "${TEST_OUTPUT}/${TEST}_" --suffix .out --idxread 1 --split_all --gzipout $DEBUG_OPT 2> ${TEST_OUTPUT}/${TEST}_error.out 1> ${TEST_OUTPUT}/${TEST}_summary.out
+EXIT_CODE=$?
+EXPECTED_EXIT_CODE=9
+check_test_results
+
+
+# Test error for special characters in sample names (utf-8)
+TEST=test_21
+echo -e "\033[1;33mSKIPPING\033[0m $TEST"
+
+
+# Test that version in barcode splitter and setup are the same (and that thus,
+# single source versioning is working)
+TEST=test_22
+echo -e "\033[1;33mSKIPPING\033[0m $TEST"
+
+
+# Test split file output
+TEST=test_23
+rm -f ${TEST_OUTPUT}/${TEST}_* 2> /dev/null
+OUTPUTS="BC1-read-1.out BC2-read-1.out BC3-read-1.out BC4-read-1.out unmatched-read-1.out summary.out error.out"
+NOT_OUTPUTS=""
+mkfifo "${TEST_DATA}/${TEST}_read1_fifo.fastq"
+cat "${TEST_DATA}/barcode_splitter1.fastq" > "${TEST_DATA}/${TEST}_read1_fifo.fastq" &
+$BARCODE_SPLITTER --mismatches 2 --bcfile "${TEST_DATA}/barcode_splitter_barcodes.txt" "${TEST_DATA}/${TEST}_read1_fifo.fastq" --prefix "${TEST_OUTPUT}/${TEST}_" --suffix .out --idxread 1 --split_all $DEBUG_OPT 2> ${TEST_OUTPUT}/${TEST}_error.out 1> ${TEST_OUTPUT}/${TEST}_summary.out
+rm "${TEST_DATA}/${TEST}_read1_fifo.fastq"
+EXIT_CODE=$?
+EXPECTED_EXIT_CODE=0
+check_test_results
+
+
+# Test duplicate barcode rows
+TEST=test_24
+rm -f ${TEST_OUTPUT}/${TEST}_* 2> /dev/null
+OUTPUTS="summary.out error.out"
+NOT_OUTPUTS="BC1-read-1.out BC2-read-1.out BC3-read-1.out BC4-read-1.out unmatched-read-1.out"
+$BARCODE_SPLITTER --bcfile "${TEST_DATA}/barcode_splitter_barcodes_dupes.txt" --prefix "${TEST_OUTPUT}/${TEST}_" "${TEST_DATA}/barcode_splitter1.fastq" "${TEST_DATA}/barcode_splitter_index.fastq" "${TEST_DATA}/barcode_splitter_index_2.fastq" --idxread 2 3 $DEBUG_OPT 2> ${TEST_OUTPUT}/${TEST}_error.out 1> ${TEST_OUTPUT}/${TEST}_summary.out
+EXIT_CODE=$?
+EXPECTED_EXIT_CODE=12
+check_test_results
+
+
+# Test gzip input and output bash process substitution
+TEST=test_25
+rm -f ${TEST_OUTPUT}/${TEST}_* 2> /dev/null
+OUTPUTS="BC1-read-1.out.gz BC2-read-1.out.gz BC3-read-1.out.gz BC4-read-1.out.gz unmatched-read-1.out.gz summary.out error.out"
+NOT_OUTPUTS=""
+CAT=cat
+OS=`uname`
+if [ "$OS" == "Darwin" ]; then
+    CAT=zcat
+fi
```

Comment:
This seems to break for me on Ubuntu (and I imagine it would break on other OSes as well).
```suggestion
CAT="gunzip -c"
```

Reply:
Yeah, I added it because it broke on my mac too and I assumed that only macs' version of cat couldn't handle compressed input, because barcode_splitter's test script uses cat on a .gz file and I knew that worked in linux.  I figured it must work on all linux systems... So does Ubuntu have an issue with this test in barcode_splitter's test script?  It must, if this test fails.

Reply:
Also, I don't know if assuming users have gunzip would work for everyone either.